File: solutions/java/kindergarten-garden/1/src/main/java/KindergartenGarden.java
import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;

class KindergartenGarden {
    // Students in alphabetical order - determines garden column assignment
    private static final List<String> STUDENTS = Arrays.asList(
        "Alice", "Bob", "Charlie", "David", "Eve", "Fred",
        "Ginny", "Harriet", "Ileana", "Joseph", "Kincaid", "Larry"
    );
    
    private final String[] rows;
    
    KindergartenGarden(String garden) {
        // Split the diagram into two rows by newline
        this.rows = garden.split("\n");
    }

    List<Plant> getPlantsOfStudent(String student) {
        // Find student's index (0 = Alice, 1 = Bob, etc.)
        int index = STUDENTS.indexOf(student);
        if (index == -1) {
            throw new IllegalArgumentException("Unknown student: " + student);
        }
        
        // Each student gets 2 consecutive cups per row, starting at column index*2
        int startCol = index * 2;
        
        List<Plant> plants = new ArrayList<>(4);
        
        // Row 0 (nearest windows): 2 plants
        plants.add(Plant.getPlant(rows[0].charAt(startCol)));
        plants.add(Plant.getPlant(rows[0].charAt(startCol + 1)));
        
        // Row 1: 2 plants
        plants.add(Plant.getPlant(rows[1].charAt(startCol)));
        plants.add(Plant.getPlant(rows[1].charAt(startCol + 1)));
        
        return plants;
    }
}

File: solutions/typescript/kindergarten-garden/1/kindergarten-garden.ts
const DEFAULT_STUDENTS: Student[] = [
  'Alice',
  'Bob',
  'Charlie',
  'David',
  'Eve',
  'Fred',
  'Ginny',
  'Harriet',
  'Ileana',
  'Joseph',
  'Kincaid',
  'Larry',
]

const PLANT_CODES = {
  G: 'grass',
  V: 'violets',
  R: 'radishes',
  C: 'clover',
} as const

type Student = string
type Plant = (typeof PLANT_CODES)[keyof typeof PLANT_CODES]
type Plants = Plant[]

export class Garden {
  private studentPlants: Map<Student, Plants>;

  constructor(diagram: string, students: Student[] = DEFAULT_STUDENTS) {
    // Sort students alphabetically
    const sortedStudents = [...students].sort();
    
    // Split diagram into rows
    const [firstRow, secondRow] = diagram.split('\n');
    
    this.studentPlants = new Map();
    
    // Each student gets 2 cups from first row and 2 from second row
    for (let i = 0; i < sortedStudents.length; i++) {
      const student = sortedStudents[i];
      const startIdx = i * 2;
      
      const plants: Plants = [
        PLANT_CODES[firstRow[startIdx] as keyof typeof PLANT_CODES],
        PLANT_CODES[firstRow[startIdx + 1] as keyof typeof PLANT_CODES],
        PLANT_CODES[secondRow[startIdx] as keyof typeof PLANT_CODES],
        PLANT_CODES[secondRow[startIdx + 1] as keyof typeof PLANT_CODES],
      ];
      
      this.studentPlants.set(student, plants);
    }
  }

  public plants(student: Student): Plants {
    const result = this.studentPlants.get(student);
    if (!result) {
      throw new Error(`Student ${student} not found`);
    }
    return result;
  }
}